pkg/agent/controller/traceflow/traceflow_controller_test.go

```diff
@@ -0,0 +1,577 @@
+						Namespace: "default",
+						Pod:       "pod-1",
+					},
+					Destination: crdv1alpha1.Destination{
+						Namespace: "default",
+						Pod:       "pod-2",
+					},
+					Packet: crdv1alpha1.Packet{
+						TransportHeader: crdv1alpha1.TransportHeader{
+							UDP: &crdv1alpha1.UDPHeader{
+								SrcPort: 90,
+								DstPort: 100,
+							},
+						},
+					},
+				},
+			},
+			expectedPacket: &binding.Packet{
+				SourceIP:        net.ParseIP(pod1IPv4),
+				SourceMAC:       pod1MAC,
+				DestinationIP:   net.ParseIP(pod2IPv4),
+				DestinationMAC:  pod2MAC,
+				IPProto:         protocol.Type_UDP,
+				SourcePort:      90,
+				DestinationPort: 100,
+				TTL:             64,
+			},
+		},
+		{
+			name: "icmp packet",
+			tf: &crdv1alpha1.Traceflow{
+				ObjectMeta: metav1.ObjectMeta{Name: "tf6", UID: "uid6"},
+				Spec: crdv1alpha1.TraceflowSpec{
+					Source: crdv1alpha1.Source{
+						Namespace: "default",
+						Pod:       "pod-1",
+					},
+					Destination: crdv1alpha1.Destination{
+						Namespace: "default",
+						Pod:       "pod-2",
+					},
+					Packet: crdv1alpha1.Packet{
+						TransportHeader: crdv1alpha1.TransportHeader{
+							ICMP: &crdv1alpha1.ICMPEchoRequestHeader{
+								ID:       10,
+								Sequence: 20,
+							},
+						},
+					},
+				},
+			},
+			expectedPacket: &binding.Packet{
+				SourceIP:       net.ParseIP(pod1IPv4),
+				SourceMAC:      pod1MAC,
+				DestinationIP:  net.ParseIP(pod2IPv4),
+				DestinationMAC: pod2MAC,
+				IPProto:        protocol.Type_ICMP,
+				ICMPType:       8,
+				ICMPEchoID:     10,
+				ICMPEchoSeq:    20,
+				TTL:            64,
+			},
+		},
+	}
+
+	for _, tt := range tcs {
+		t.Run(tt.name, func(t *testing.T) {
+			tfc := newFakeTraceflowController(t, []runtime.Object{tt.tf}, nil, nil, nil)
+			defer tfc.mockController.Finish()
+
+			podInterfaces := tfc.interfaceStore.GetContainerInterfacesByPod("pod-1", "default")
+
+			pkt, err := tfc.preparePacket(tt.tf, podInterfaces[0], tt.receiverOnly)
+			if tt.expectedErr == "" {
+				require.NoError(t, err)
+				assert.Equal(t, tt.expectedPacket, pkt)
+			} else {
+				require.EqualError(t, err, tt.expectedErr)
+			}
+		})
+	}
+}
+
+func TestErrTraceflowCRD(t *testing.T) {
+	tf := &crdv1alpha1.Traceflow{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "dummy-traceflow",
+			UID:  "uid",
+		},
+		Spec: crdv1alpha1.TraceflowSpec{
+			Source: crdv1alpha1.Source{
+				Namespace: "default",
+				Pod:       "pod-1",
+			},
+			Destination: crdv1alpha1.Destination{
+				Namespace: "default",
+				Pod:       "pod-2",
+			},
+		},
+		Status: crdv1alpha1.TraceflowStatus{
+			Phase:        crdv1alpha1.Running,
+			DataplaneTag: 1,
+		},
+	}
+	expectedTf := tf
+	reason := "failed"
+	expectedTf.Status.Phase = crdv1alpha1.Failed
+	expectedTf.Status.Reason = reason
+
+	tfc := newFakeTraceflowController(t, []runtime.Object{tf}, nil, nil, nil)
+	defer tfc.mockController.Finish()
+
+	gotTf, err := tfc.errorTraceflowCRD(tf, reason)
```

Comment:
instead of gotTf, you can use processedTraceflow, or tfReceived, just a suggestion.